panic on nil access on getting statefulset status

When a StatefulSet is unable to initialize an initial deployment because of some reason like resource quota it never gets an observed generation. The check at [resourceKinds.go:212](https://github.com/weaveworks/flux/blob/1.3.0/cluster/kubernetes/resourcekinds.go#L212) gets a nil dereference.

~~~
ts=2018-05-01T01:20:23.152400197Z caller=main.go:422 addr=:3030
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xf85f1a]

goroutine 82 [running]:
github.com/weaveworks/flux/cluster/kubernetes.makeStatefulSetPodController(0xc421549900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/weaveworks/flux/cluster/kubernetes/resourcekinds.go:212 +0xea
github.com/weaveworks/flux/cluster/kubernetes.(*statefulSetKind).getPodControllers(0x1b6fbb0, 0xc420226960, 0xc4201670a0, 0x14, 0x1, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/weaveworks/flux/cluster/kubernetes/resourcekinds.go:203 +0x13d
github.com/weaveworks/flux/cluster/kubernetes.(*Cluster).ImagesToFetch(0xc420226960, 0xc42004e380)
        /go/src/github.com/weaveworks/flux/cluster/kubernetes/kubernetes.go:390 +0x248
github.com/weaveworks/flux/cluster/kubernetes.(*Cluster).ImagesToFetch-fm(0xdf8475800)
        /go/src/github.com/weaveworks/flux/cmd/fluxd/main.go:242 +0x2a
github.com/weaveworks/flux/registry/cache.(*Warmer).Loop(0xc4203a9f00, 0x1ad49e0, 0xc4203fe0c0, 0xc42012ee40, 0xc4203cc590, 0xc4203cc110)
        /go/src/github.com/weaveworks/flux/registry/cache/warming.go:55 +0x89
created by main.main
        /go/src/github.com/weaveworks/flux/cmd/fluxd/main.go:415 +0x423e
~~~

The statefulset in question that is crashing flux looks like has a status section like:

~~~json
    "status": {
      "replicas": 0
    }
~~~

It is currently stuck since the user has used all of their current quota.

--- a/cluster/kubernetes/resourcekinds.go
+++ b/cluster/kubernetes/resourcekinds.go
@@ -209,7 +209,8 @@ func (dk *statefulSetKind) getPodControllers(c *Cluster, namespace string) ([]po
 func makeStatefulSetPodController(statefulSet *apiapps.StatefulSet) podController {
 	var status string
 	objectMeta, statefulSetStatus := statefulSet.ObjectMeta, statefulSet.Status
-	if *statefulSetStatus.ObservedGeneration >= objectMeta.Generation {
+	// The type of ObservedGeneration is *int64, unlike other controllers.
+	if statefulSetStatus.ObservedGeneration != nil && *statefulSetStatus.ObservedGeneration >= objectMeta.Generation {
 		// the definition has been updated; now let's see about the replicas
 		updated, wanted := statefulSetStatus.UpdatedReplicas, *statefulSet.Spec.Replicas
 		if updated == wanted {

